packages/Search/src/details/DTK_DetailsDistributor.hpp

```diff
@@ -0,0 +1,246 @@
+
+#include <mpi.h>
+
+#include <algorithm>
+#include <cassert>
+#include <vector>
+
+#define REORDER_RECV YES
+
+namespace DataTransferKit
+{
+namespace Details
+{
+
+// clang-format off
+// copy/paste from https://github.com/kokkos/kokkos-tutorials/blob/ee619447cee9e57b831e34b151d0868958b6b1e5/Intro-Full/Exercises/mpi_exch/mpi_exch.cpp
+// * added Experimental nested namespace qualifier for Max
+// * declared function static to avoid multiple definition error at link time
+// * return early if input argument is empty
+// * replace parallel_reduce with std::max_element because I couldn't get it to compile
+// * append size to offsets in main driver but might want to handle it here
+static void extract_and_sort_ranks(
+    Kokkos::View<int*, Kokkos::HostSpace> destination_ranks,
+    Kokkos::View<int*, Kokkos::HostSpace> permutation,
+    std::vector<int>& unique_ranks,
+    std::vector<int>& offsets,
+    std::vector<int>& counts) {
+  int mpi_rank;
+  MPI_Comm_rank(MPI_COMM_WORLD, &mpi_rank);
+  auto n = destination_ranks.extent(0);
+if ( n == 0 ) return;
+  using ST = decltype(n);
+  Kokkos::View<int*, Kokkos::HostSpace> tmp_ranks("tmp ranks", destination_ranks.extent(0));
+  Kokkos::deep_copy(tmp_ranks, destination_ranks);
+  int offset = 0;
+  // this implements a "sort" which is O(N * R) where (R) is
+  // the total number of unique destination ranks.
+  // it performs better than other algorithms in
+  // the case when (R) is small, but results may vary
+  while (true) {
+    int next_biggest_rank;
+    //Kokkos::parallel_reduce("find next biggest rank", n, KOKKOS_LAMBDA(ST i, int& local_max) {
+    //  auto r = tmp_ranks(i);
+    //  local_max = (r > local_max) ? r : local_max;
+    //}, Kokkos::Experimental::Max<int, Kokkos::HostSpace>(next_biggest_rank));
+    next_biggest_rank = *std::max_element(tmp_ranks.data(), tmp_ranks.data() + n);
+    if (next_biggest_rank == -1) break;
+    unique_ranks.push_back(next_biggest_rank);
+    offsets.push_back(offset);
+    Kokkos::View<int, Kokkos::HostSpace> total("total");
+    Kokkos::parallel_scan("process biggest rank items", Kokkos::RangePolicy<Kokkos::Serial>(0, n),
+    KOKKOS_LAMBDA(ST i, int& index, const bool last_pass) {
+      if (last_pass && (tmp_ranks(i) == next_biggest_rank)) {
+        permutation(i) = index + offset;
+      }
+      if (tmp_ranks(i) == next_biggest_rank) ++index;
+      if (last_pass) {
+        if (i + 1 == tmp_ranks.extent(0)) {
+          total() = index;
+        }
+        if (tmp_ranks(i) == next_biggest_rank) {
+          tmp_ranks(i) = -1;
+        }
+      }
+    });
+    auto host_total = Kokkos::create_mirror_view_and_copy(
+        Kokkos::HostSpace(), total);
+    auto count = host_total();
+    counts.push_back(count);
+    offset += count;
+  }
+}
+// clang-format on
+
+class Distributor
+{
+  public:
+    Distributor( MPI_Comm comm )
+        : _comm( comm )
+    {
+    }
+
+    template <typename T>
+    static T *notNullPtr( T *p )
```

Comment:
What's the point of that function?

Reply:
I need to think about the name a bit more carefully and comment but in short, OpenMPI routines assert that some of the pointers passed as argument are not null even when no data will be communicated (`std::vector::data()` returns `nullptr` when the container is empty)

Reply:
I added a note above the function declaration to capture this.